model/map/src/main/java/org/keycloak/models/map/storage/MapFieldPredicates.java

```diff
@@ -264,6 +298,71 @@ private static String ensureEqSingleValue(SearchableModelField<?> field, String
     private static MapModelCriteriaBuilder<Object, AbstractRealmEntity<Object>, RealmModel> checkRealmsWithComponentType(MapModelCriteriaBuilder<Object, AbstractRealmEntity<Object>, RealmModel> mcb, Operator op, Object[] values) {
         String providerType = ensureEqSingleValue(RealmModel.SearchableFields.COMPONENT_PROVIDER_TYPE, "component_provider_type", op, values);
         Function<AbstractRealmEntity<Object>, ?> getter = realmEntity -> realmEntity.getComponents().anyMatch(component -> component.getProviderType().equals(providerType));
+        return mcb.fieldCompare(Boolean.TRUE::equals, getter);
+    }
+
+    private static MapModelCriteriaBuilder<Object, AbstractUserSessionEntity<Object>, UserSessionModel> checkUserSessionContainsAuthenticatedClientSession(MapModelCriteriaBuilder<Object, AbstractUserSessionEntity<Object>, UserSessionModel> mcb, Operator op, Object[] values) {
+        String clientId = ensureEqSingleValue(UserSessionModel.SearchableFields.CLIENT_ID, "client_id", op, values);
+        Function<AbstractUserSessionEntity<Object>, ?> getter;
+        getter = use -> (use.getAuthenticatedClientSessions().containsKey(clientId));
+
+        return mcb.fieldCompare(Boolean.TRUE::equals, getter);
+    }
+
+    private static MapModelCriteriaBuilder<Object, AbstractUserSessionEntity<Object>, UserSessionModel> isUserSessionExpired(MapModelCriteriaBuilder<Object, AbstractUserSessionEntity<Object>, UserSessionModel> mcb, Operator op, Object[] values) {
+        if (op != Operator.EQ) {
+            throw new CriterionNotSupportedException(UserSessionModel.SearchableFields.IS_EXPIRED, op);
+        }
+        if (values == null || values.length != 5) {
+            throw new CriterionNotSupportedException(UserSessionModel.SearchableFields.IS_EXPIRED, op, "Invalid arguments, expected (expiredRememberMe, expiredRefreshRememberMe, expired, expiredRefresh, expiredOffline), got: " + Arrays.toString(values));
+        }
+
+        int expiredRememberMe = (int) values[0];
+        int expiredRefreshRememberMe = (int) values[1];
+        int expired = (int) values[2];
+        int expiredRefresh = (int) values[3];
+        int expiredOffline = (int) values[4];
+
+        Function<AbstractUserSessionEntity<Object>, ?> getter;
+        getter = use -> {
```

Comment:
Even though this is really nice addition to MapStorage, do we have some plan for how we will make this work for JPA-MAP storage?